MovieTime/MovieTime/Service/MTMovieDBService.swift

```diff
@@ -0,0 +1,87 @@
+//
+//  MTMovieDBService.swift
+//  MovieTime
+//
+//  Created by Enrique Melgarejo on 08/01/18.
+//  Copyright © 2018 Choynowski. All rights reserved.
+//
+
+import Foundation
+
+protocol MTMovieDBServiceProtocol: class {
+    init(service: MTServiceProtocol, helper: MTServiceHelperProtocol)
+
+    func requestConfiguration(completion: ((Data?, MTNetworkError?) -> Void)?)
+    func requestGenresList(completion: ((Data?, MTNetworkError?) -> Void)?)
+    func requestUpcomingMovies(forPage page: Int, completion: ((Data?, MTNetworkError?) -> Void)?)
+    func requestMovieDetail(forId movieId: Int, completion: ((Data?, MTNetworkError?) -> Void)?)
+}
+
+final class MTMovieDBService: MTMovieDBServiceProtocol {
+
+    private let urlServer = URL(string: "https://api.themoviedb.org/3")
+    private let currentService: MTServiceProtocol
+    private let serviceHelper: MTServiceHelperProtocol
+
+    private let apiKey = "1f54bd990f1cdfb230adb312546d765d"
+    private let language = "en-US"
+
+    init(service: MTServiceProtocol = MTService(), helper: MTServiceHelperProtocol = MTServiceHelper()) {
+        currentService = service
+        serviceHelper = helper
+    }
+
+    func requestConfiguration(completion: ((Data?, MTNetworkError?) -> Void)?) {
+
+        let endpoint = "/configuration"
+        let parameters: MTParameters = ["api_key": apiKey,
+                                        "language": language]
+        let url = urlServer?.appendingPathComponent(endpoint)
+
+        executeRequest(url: url, method: .get, parameters: parameters, completion: completion)
+    }
+
+    func requestGenresList(completion: ((Data?, MTNetworkError?) -> Void)?) {
+
+        let endpoint = "/genre/movie/list"
+        let parameters: MTParameters = ["api_key": apiKey,
+                                        "language": language]
+        let url = urlServer?.appendingPathComponent(endpoint)
+
+        executeRequest(url: url, method: .get, parameters: parameters, completion: completion)
+    }
+
+    func requestUpcomingMovies(forPage page: Int = 1, completion: ((Data?, MTNetworkError?) -> Void)?) {
```

Comment:
if completion is optional, does it make sense to force users from this API to always have to pass a `nil` value if they don't want any completion action? Is there a way to improve that?

Reply:
Since these are `GET` requests, so they will return some information. Using a default parameter value would not fit into the implementation context.
It is a optional to make it easier to use the method (when passing a block with weak reference for example).
But it could be a non-option block, since performing the request without a return is of no use.

Reply:
🎉  I agree a non-optional block would better match the requirements for this method